// webgl/lib/webgl/constants.js
require( './globals' );
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
module.exports = {
	bufferbit:
	{
		depth_buffer_bit: WEBGL2.DEPTH_BUFFER_BIT,
		stencil_buffer_bit: WEBGL2.STENCIL_BUFFER_BIT,
		color_buffer_bit: WEBGL2.COLOR_BUFFER_BIT,
	},
	rednering_primitives:
	{
		points: WEBGL2.POINTS,
		lines: WEBGL2.LINES,
		line_loop: WEBGL2.LINE_LOOP,
		line_strip: WEBGL2.LINE_STRIP,
		triangles: WEBGL2.TRIANGLES,
		triangle_strip: WEBGL2.TRIANGLE_STRIP,
		triangle_fan: WEBGL2.TRIANGLE_FAN,
	},
	blend:
	{
		mode:
		{
			zero: WEBGL2.ZERO,
			one: WEBGL2.ONE,
			src_color: WEBGL2.SRC_COLOR,
			one_minus_src_color: WEBGL2.ONE_MINUS_SRC_COLOR,
			src_alpha: WEBGL2.SRC_ALPHA,
			one_minus_src_alpha: WEBGL2.ONE_MINUS_SRC_ALPHA,
			dst_alpha: WEBGL2.DST_ALPHA,
			one_minus_dst_alpha: WEBGL2.ONE_MINUS_DST_ALPHA,
			dst_color: WEBGL2.DST_COLOR,
			one_minus_dst_color: WEBGL2.ONE_MINUS_DST_COLOR,
			src_alpha_saturate: WEBGL2.SRC_ALPHA_SATURATE,
			constant_color: WEBGL2.CONSTANT_COLOR,
			one_minus_constant_color: WEBGL2.ONE_MINUS_CONSTANT_COLOR,
			constant_alpha: WEBGL2.CONSTANT_ALPHA,
			one_minus_constant_alpha: WEBGL2.ONE_MINUS_CONSTANT_ALPHA,
		},
		equation:
		{
			add: WEBGL2.FUNC_ADD,
			substract: WEBGL2.FUNC_SUBSTRACT,
			reverse_subtract: WEBGL2.FUNC_REVERSE_SUBTRACT,
		}
	},
	param_info:
	{
		blend_equation: WEBGL2.BLEND_EQUATION,
		blend_equation_rgb: WEBGL2.BLEND_EQUATION_RGB,
		blend_equation_alpha: WEBGL2.BLEND_EQUATION_ALPHA,
		blend_dst_rgb: WEBGL2.BLEND_DST_RGB,
		blend_src_rgb: WEBGL2.BLEND_SRC_RGB,
		blend_dst_alpha: WEBGL2.BLEND_DST_ALPHA,
		blend_src_alpha: WEBGL2.BLEND_SRC_ALPHA,
		blend_color: WEBGL2.BLEND_COLOR,
		array_buffer_binding: WEBGL2.ARRAY_BUFFER_BINDING,
		element_array_buffer_binding: WEBGL2.ELEMENT_ARRAY_BUFFER_BINDING,
		line_width: WEBGL2.LINE_WIDTH,
		aliased_point_size_range: WEBGL2.ALIASED_POINT_SIZE_RANGE,
		aliased_line_width_range: WEBGL2.ALIASED_LINE_WIDTH_RANGE,
		cull_face_mode: WEBGL2.CULL_FACE_MODE,
		front_face: WEBGL2.FRONT_FACE,
		depth_range: WEBGL2.DEPTH_RANGE,
		depth_writemask: WEBGL2.DEPTH_WRITEMASK,
		depth_clear_value: WEBGL2.DEPTH_CLEAR_VALUE,
		depth_func: WEBGL2.DEPTH_FUNC,
		stencil_clear_value: WEBGL2.STENCIL_CLEAR_VALUE,
		stencil_func: WEBGL2.STENCIL_FUNC,
		stencil_fail: WEBGL2.STENCIL_FAIL,
		stencil_pass_depth_fail: WEBGL2.STENCIL_PASS_DEPTH_FAIL,
		stencil_pass_depth_pass: WEBGL2.STENCIL_PASS_DEPTH_PASS,
		stencil_ref: WEBGL2.STENCIL_REF,
		stencil_value_mask: WEBGL2.STENCIL_VALUE_MASK,
		stencil_writemask: WEBGL2.STENCIL_WRITEMASK,
		stencil_back_func: WEBGL2.STENCIL_BACK_FUNC,
		stencil_back_fail: WEBGL2.STENCIL_BACK_FAIL,
		stencil_back_pass_depth_fail: WEBGL2.STENCIL_BACK_PASS_DEPTH_FAIL,
		stencil_back_pass_depth_pass: WEBGL2.STENCIL_BACK_PASS_DEPTH_PASS,
		stencil_back_ref: WEBGL2.STENCIL_BACK_REF,
		stencil_back_value_mask: WEBGL2.STENCIL_BACK_VALUE_MASK,
		stencil_back_writemask: WEBGL2.STENCIL_BACK_WRITEMASK,
		viewport: WEBGL2.VIEWPORT,
		scissor_box: WEBGL2.SCISSOR_BOX,
		color_clear_value: WEBGL2.COLOR_CLEAR_VALUE,
		color_writemask: WEBGL2.COLOR_WRITEMASK,
		unpack_alignment: WEBGL2.UNPACK_ALIGNMENT,
		pack_alignment: WEBGL2.PACK_ALIGNMENT,
		max_texture_size: WEBGL2.MAX_TEXTURE_SIZE,
		max_viewport_dims: WEBGL2.MAX_VIEWPORT_DIMS,
		subpixel_bits: WEBGL2.SUBPIXEL_BITS,
		red_bits: WEBGL2.RED_BITS,
		green_bits: WEBGL2.GREEN_BITS,
		blue_bits: WEBGL2.BLUE_BITS,
		alpha_bits: WEBGL2.ALPHA_BITS,
		depth_bits: WEBGL2.DEPTH_BITS,
		stencil_bits: WEBGL2.STENCIL_BITS,
		polygon_offset_units: WEBGL2.POLYGON_OFFSET_UNITS,
		polygon_offset_factor: WEBGL2.POLYGON_OFFSET_FACTOR,
		texture_binding_2d: WEBGL2.TEXTURE_BINDING_2D,
		sample_buffers: WEBGL2.SAMPLE_BUFFERS,
		samples: WEBGL2.SAMPLES,
		sample_coverage_value: WEBGL2.SAMPLE_COVERAGE_VALUE,
		sample_coverage_invert: WEBGL2.SAMPLE_COVERAGE_INVERT,
		compressed_texture_formats: WEBGL2.COMPRESSED_TEXTURE_FORMATS,
		vendor: WEBGL2.VENDOR,
		renderer: WEBGL2.RENDERER,
		version: WEBGL2.VERSION,
		implementation_color_read_type: WEBGL2.IMPLEMENTATION_COLOR_READ_TYPE,
		implementation_color_read_format: WEBGL2.IMPLEMENTATION_COLOR_READ_FORMAT,
		browser_default_webgl: WEBGL2.BROWSER_DEFAULT_WEBGL,
		read_buffer: WEBGL2.READ_BUFFER,
		unpack_row_length: WEBGL2.UNPACK_ROW_LENGTH,
		unpack_skip_rows: WEBGL2.UNPACK_SKIP_ROWS,
		unpack_skip_pixels: WEBGL2.UNPACK_SKIP_PIXELS,
		pack_row_length: WEBGL2.PACK_ROW_LENGTH,
		pack_skip_rows: WEBGL2.PACK_SKIP_ROWS,
		pack_skip_pixels: WEBGL2.PACK_SKIP_PIXELS,
		texture_binding_3d: WEBGL2.TEXTURE_BINDING_3D,
		unpack_skip_images: WEBGL2.UNPACK_SKIP_IMAGES,
		unpack_image_height: WEBGL2.UNPACK_IMAGE_HEIGHT,
		max_3d_texture_size: WEBGL2.MAX_3D_TEXTURE_SIZE,
		max_elements_vertices: WEBGL2.MAX_ELEMENTS_VERTICES,
		max_elements_indices: WEBGL2.MAX_ELEMENTS_INDICES,
		max_texture_lod_bias: WEBGL2.MAX_TEXTURE_LOD_BIAS,
		max_fragment_uniform_components: WEBGL2.MAX_FRAGMENT_UNIFORM_COMPONENTS,
		max_vertex_uniform_components: WEBGL2.MAX_VERTEX_UNIFORM_COMPONENTS,
		max_array_texture_layers: WEBGL2.MAX_ARRAY_TEXTURE_LAYERS,
		min_program_texel_offset: WEBGL2.MIN_PROGRAM_TEXEL_OFFSET,
		max_program_texel_offset: WEBGL2.MAX_PROGRAM_TEXEL_OFFSET,
		max_varying_components: WEBGL2.MAX_VARYING_COMPONENTS,
		fragment_shader_derivative_hint: WEBGL2.FRAGMENT_SHADER_DERIVATIVE_HINT,
		rasterizer_discard: WEBGL2.RASTERIZER_DISCARD,
		vertex_array_binding: WEBGL2.VERTEX_ARRAY_BINDING,
		max_vertex_output_components: WEBGL2.MAX_VERTEX_OUTPUT_COMPONENTS,
		max_fragment_input_components: WEBGL2.MAX_FRAGMENT_INPUT_COMPONENTS,
		max_server_wait_timeout: WEBGL2.MAX_SERVER_WAIT_TIMEOUT,
		max_element_index: WEBGL2.MAX_ELEMENT_INDEX,
	},
	buffer:
	{
		target:
		{
			array: WEBGL2.ARRAY_BUFFER,
			element_array: WEBGL2.ELEMENT_ARRAY_BUFFER,
			copy:
			{
				read: WEBGL2.COPY_READ_BUFFER,
				write: WEBGL2.COPY_WRITE_BUFFER,
			},
			pixel:
			{
				pack: WEBGL2.PIXEL_PACK_BUFFER,
				unpack: WEBGL2.PIXEL_UNPACK_BUFFER,
			}
		},
		use:
		{
			static_draw: WEBGL2.STATIC_DRAW,
			stream_draw: WEBGL2.STREAM_DRAW,
			dynamic_draw: WEBGL2.DYNAMIC_DRAW,
		},
		size: WEBGL2.BUFFER_SIZE,
		usage: WEBGL2.BUFFER_USAGE,
		binding:
		{
			pixel:
			{
				pack: WEBGL2.PIXEL_PACK_BUFFER_BINDING,
				unpack: WEBGL2.PIXEL_UNPACK_BUFFER_BINDING,
			},
			copy:
			{
				read: WEBGL2.COPY_READ_BUFFER_BINDING,
				write: WEBGL2.COPY_WRITE_BUFFER_BINDING,
			},
		},
	},
	vertex_attributes:
	{
		current_vertex_attrib: WEBGL2.CURRENT_VERTEX_ATTRIB,
		vertex_attrib_array_enabled: WEBGL2.VERTEX_ATTRIB_ARRAY_ENABLED,
		vertex_attrib_array_size: WEBGL2.VERTEX_ATTRIB_ARRAY_SIZE,
		vertex_attrib_array_stride: WEBGL2.VERTEX_ATTRIB_ARRAY_STRIDE,
		vertex_attrib_array_type: WEBGL2.VERTEX_ATTRIB_ARRAY_TYPE,
		vertex_attrib_array_normalized: WEBGL2.VERTEX_ATTRIB_ARRAY_NORMALIZED,
		vertex_attrib_array_pointer: WEBGL2.VERTEX_ATTRIB_ARRAY_POINTER,
		vertex_attrib_array_buffer_binding: WEBGL2.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING,
		vertex_attrib_array_integer: WEBGL2.VERTEX_ATTRIB_ARRAY_INTEGER,
		vertex_attrib_array_divisor: WEBGL2.VERTEX_ATTRIB_ARRAY_DIVISOR,
	},
	culling:
	{
		face: WEBGL2.CULL_FACE,
		front: WEBGL2.FRONT,
		back: WEBGL2.BACK,
		front_and_back: WEBGL2.FRONT_AND_BACK,
	},
	enable_disable:
	{
		blend: WEBGL2.BLEND,
		depth_test: WEBGL2.DEPTH_TEST,
		dither: WEBGL2.DITHER,
		polygon_offset_fill: WEBGL2.POLYGON_OFFSET_FILL,
		sample_alpha_to_coverage: WEBGL2.SAMPLE_ALPHA_TO_COVERAGE,
		sample_coverage: WEBGL2.SAMPLE_COVERAGE,
		scissor_test: WEBGL2.SCISSOR_TEST,
		stencil_test: WEBGL2.STENCIL_TEST,
	},
	error:
	{
		none: WEBGL2.NO_ERROR,
		invalid:
		{
			enum: WEBGL2.INVALID_ENUM,
			value: WEBGL2.INVALID_VALUE,
			operation: WEBGL2.INVALID_OPERATION,
		},
		out_of_memory: WEBGL2.OUT_OF_MEMORY,
		context_lost_webgl: WEBGL2.CONTEXT_LOST_WEBGL,
	},
	winding:
	{
		cw: WEBGL2.CW,
		ccw: WEBGL2.CCW,
	},
	hint:
	{
		dont_care: WEBGL2.DONT_CARE,
		fastest: WEBGL2.FASTEST,
		nicest: WEBGL2.NICEST,
		generate_mipmap_hint: WEBGL2.GENERATE_MIPMAP_HINT,
	},
	types:
	{
		byte: WEBGL2.BYTE,
		ubyte: WEBGL2.UNSIGNED_BYTE,
		short: WEBGL2.SHORT,
		ushort: WEBGL2.UNSIGNED_SHORT,
		int: WEBGL2.INT,
		uint: WEBGL2.UNSIGNED_INT,
		float: WEBGL2.FLOAT,
		float_mat2x3: WEBGL2.FLOAT_MAT2x3,
		float_mat2x4: WEBGL2.FLOAT_MAT2x4,
		float_mat3x2: WEBGL2.FLOAT_MAT3x2,
		float_mat3x4: WEBGL2.FLOAT_MAT3x4,
		float_mat4x2: WEBGL2.FLOAT_MAT4x2,
		float_mat4x3: WEBGL2.FLOAT_MAT4x3,
		uint_vec2: WEBGL2.UNSIGNED_INT_VEC2,
		uint_vec3: WEBGL2.UNSIGNED_INT_VEC3,
		uint_vec4: WEBGL2.UNSIGNED_INT_VEC4,
		unsigned_normalized: WEBGL2.UNSIGNED_NORMALIZED,
		signed_normalized: WEBGL2.SIGNED_NORMALIZED,
	},
	pixel:
	{
		format:
		{
			depth_component: WEBGL2.DEPTH_COMPONENT,
			alpha: WEBGL2.ALPHA,
			rgb: WEBGL2.RGB,
			rgba: WEBGL2.RGBA,
			luminance: WEBGL2.LUMINANCE,
			luminance_alpha: WEBGL2.LUMINANCE_ALPHA,
		},
		type:
		{
			ubyte: WEBGL2.UNSIGNED_BYTE,
			ushort_4_4_4_4: WEBGL2.UNSIGNED_SHORT_4_4_4_4,
			ushort_5_5_5_1: WEBGL2.UNSIGNED_SHORT_5_5_5_1,
			ushort_5_6_5: WEBGL2.UNSIGNED_SHORT_5_6_5,
			uint_2_10_10_10_rev: WEBGL2.UNSIGNED_INT_2_10_10_10_REV,
			uint_10f_11f_11f_rev: WEBGL2.UNSIGNED_INT_10F_11F_11F_REV,
			uint_5_9_9_9_rev: WEBGL2.UNSIGNED_INT_5_9_9_9_REV,
			float_32_uint_24_8_rev: WEBGL2.FLOAT_32_UNSIGNED_INT_24_8_REV,
			uint_24_8: WEBGL2.UNSIGNED_INT_24_8,
			half_float: WEBGL2.HALF_FLOAT,
			rg: WEBGL2.RG,
			rg_integer: WEBGL2.RG_INTEGER,
			int_2_10_10_10_rev: WEBGL2.INT_2_10_10_10_REV,
		},
		storage:
		{
			unpack_flip_y_webgl: WEBGL2.UNPACK_FLIP_Y_WEBGL,
			unpack_premultiply_alpha_webgl: WEBGL2.UNPACK_PREMULTIPLY_ALPHA_WEBGL,
			unpack_colorspace_conversion_webgl: WEBGL2.UNPACK_COLORSPACE_CONVERSION_WEBGL,
		},
	},
	shader:
	{
		type:
		{
			fragment_shader: WEBGL2.FRAGMENT_SHADER,
			vertex_shader: WEBGL2.VERTEX_SHADER,
		},
		compile_status: WEBGL2.COMPILE_STATUS,
		delete_status: WEBGL2.DELETE_STATUS,
		link_status: WEBGL2.LINK_STATUS,
		validate_status: WEBGL2.VALIDATE_STATUS,
		attached_shaders: WEBGL2.ATTACHED_SHADERS,
		active_attributes: WEBGL2.ACTIVE_ATTRIBUTES,
		active_uniforms: WEBGL2.ACTIVE_UNIFORMS,
		max_vertex_attribs: WEBGL2.MAX_VERTEX_ATTRIBS,
		max_vertex_uniform_vectors: WEBGL2.MAX_VERTEX_UNIFORM_VECTORS,
		max_varying_vectors: WEBGL2.MAX_VARYING_VECTORS,
		max_combined_texture_image_units: WEBGL2.MAX_COMBINED_TEXTURE_IMAGE_UNITS,
		max_vertex_texture_image_units: WEBGL2.MAX_VERTEX_TEXTURE_IMAGE_UNITS,
		max_texture_image_units: WEBGL2.MAX_TEXTURE_IMAGE_UNITS,
		max_fragment_uniform_vectors: WEBGL2.MAX_FRAGMENT_UNIFORM_VECTORS,
		shader_type: WEBGL2.SHADER_TYPE,
		shading_language_version: WEBGL2.SHADING_LANGUAGE_VERSION,
		current_program: WEBGL2.CURRENT_PROGRAM,

		precision:
		{
			low_float: WEBGL2.LOW_FLOAT,
			medium_float: WEBGL2.MEDIUM_FLOAT,
			high_float: WEBGL2.HIGH_FLOAT,
			low_int: WEBGL2.LOW_INT,
			medium_int: WEBGL2.MEDIUM_INT,
			high_int: WEBGL2.HIGH_INT,
		},
	},
	depth:
	{
		test:
		{
			never: WEBGL2.NEVER,
			always: WEBGL2.ALWAYS,
			less: WEBGL2.LESS,
			equal: WEBGL2.EQUAL,
			lequal: WEBGL2.LEQUAL,
			greater: WEBGL2.GREATER,
			gequal: WEBGL2.GEQUAL,
			notequal: WEBGL2.NOTEQUAL,
		},
	},
	stencil:
	{
		test:
		{
			never: WEBGL2.NEVER,
			always: WEBGL2.ALWAYS,
			less: WEBGL2.LESS,
			equal: WEBGL2.EQUAL,
			lequal: WEBGL2.LEQUAL,
			greater: WEBGL2.GREATER,
			gequal: WEBGL2.GEQUAL,
			notequal: WEBGL2.NOTEQUAL,
		},
		action:
		{
			keep: WEBGL2.KEEP,
			replace: WEBGL2.REPLACE,
			incr: WEBGL2.INCR,
			decr: WEBGL2.DECR,
			invert: WEBGL2.INVERT,
			incr_wrap: WEBGL2.INCR_WRAP,
			decr_wrap: WEBGL2.DECR_WRAP,
		},
	},
	textures:
	{
		nearest: WEBGL2.NEAREST,
		linear: WEBGL2.LINEAR,
		nearest_mipmap_nearest: WEBGL2.NEAREST_MIPMAP_NEAREST,
		linear_mipmap_nearest: WEBGL2.LINEAR_MIPMAP_NEAREST,
		nearest_mipmap_linear: WEBGL2.NEAREST_MIPMAP_LINEAR,
		linear_mipmap_linear: WEBGL2.LINEAR_MIPMAP_LINEAR,
		texture_mag_filter: WEBGL2.TEXTURE_MAG_FILTER,
		texture_min_filter: WEBGL2.TEXTURE_MIN_FILTER,
		texture_wrap_s: WEBGL2.TEXTURE_WRAP_S,
		texture_wrap_t: WEBGL2.TEXTURE_WRAP_T,
		texture_2d: WEBGL2.TEXTURE_2D,
		texture: WEBGL2.TEXTURE,
		texture_cube_map: WEBGL2.TEXTURE_CUBE_MAP,
		texture_binding_cube_map: WEBGL2.TEXTURE_BINDING_CUBE_MAP,
		texture_cube_map_positive_x: WEBGL2.TEXTURE_CUBE_MAP_POSITIVE_X,
		texture_cube_map_negative_x: WEBGL2.TEXTURE_CUBE_MAP_NEGATIVE_X,
		texture_cube_map_positive_y: WEBGL2.TEXTURE_CUBE_MAP_POSITIVE_Y,
		texture_cube_map_negative_y: WEBGL2.TEXTURE_CUBE_MAP_NEGATIVE_Y,
		texture_cube_map_positive_z: WEBGL2.TEXTURE_CUBE_MAP_POSITIVE_Z,
		texture_cube_map_negative_z: WEBGL2.TEXTURE_CUBE_MAP_NEGATIVE_Z,
		max_cube_map_texture_size: WEBGL2.MAX_CUBE_MAP_TEXTURE_SIZE,
		active_texture: WEBGL2.ACTIVE_TEXTURE,
		repeat: WEBGL2.REPEAT,
		clamp_to_edge: WEBGL2.CLAMP_TO_EDGE,
		mirrored_repeat: WEBGL2.MIRRORED_REPEAT,
		red: WEBGL2.RED,
		rgb8: WEBGL2.RGB8,
		rgba8: WEBGL2.RGBA8,
		rgb10_a2: WEBGL2.RGB10_A2,
		texture_3d: WEBGL2.TEXTURE_3D,
		texture_wrap_r: WEBGL2.TEXTURE_WRAP_R,
		texture_min_lod: WEBGL2.TEXTURE_MIN_LOD,
		texture_max_lod: WEBGL2.TEXTURE_MAX_LOD,
		texture_base_level: WEBGL2.TEXTURE_BASE_LEVEL,
		texture_max_level: WEBGL2.TEXTURE_MAX_LEVEL,
		texture_compare_mode: WEBGL2.TEXTURE_COMPARE_MODE,
		texture_compare_func: WEBGL2.TEXTURE_COMPARE_FUNC,
		srgb: WEBGL2.SRGB,
		srgb8: WEBGL2.SRGB8,
		srgb8_alpha8: WEBGL2.SRGB8_ALPHA8,
		compare_ref_to_texture: WEBGL2.COMPARE_REF_TO_TEXTURE,
		rgba32f: WEBGL2.RGBA32F,
		rgb32f: WEBGL2.RGB32F,
		rgba16f: WEBGL2.RGBA16F,
		rgb16f: WEBGL2.RGB16F,
		texture_2d_array: WEBGL2.TEXTURE_2D_ARRAY,
		texture_binding_2d_array: WEBGL2.TEXTURE_BINDING_2D_ARRAY,
		r11f_g11f_b10f: WEBGL2.R11F_G11F_B10F,
		rgb9_e5: WEBGL2.RGB9_E5,
		rgba32ui: WEBGL2.RGBA32UI,
		rgb32ui: WEBGL2.RGB32UI,
		rgba16ui: WEBGL2.RGBA16UI,
		rgb16ui: WEBGL2.RGB16UI,
		rgba8ui: WEBGL2.RGBA8UI,
		rgb8ui: WEBGL2.RGB8UI,
		rgba32i: WEBGL2.RGBA32I,
		rgb32i: WEBGL2.RGB32I,
		rgba16i: WEBGL2.RGBA16I,
		rgb16i: WEBGL2.RGB16I,
		rgba8i: WEBGL2.RGBA8I,
		rgb8i: WEBGL2.RGB8I,
		red_integer: WEBGL2.RED_INTEGER,
		rgb_integer: WEBGL2.RGB_INTEGER,
		rgba_integer: WEBGL2.RGBA_INTEGER,
		r8: WEBGL2.R8,
		rg8: WEBGL2.RG8,
		r16f: WEBGL2.R16F,
		r32f: WEBGL2.R32F,
		rg16f: WEBGL2.RG16F,
		rg32f: WEBGL2.RG32F,
		r8i: WEBGL2.R8I,
		r8ui: WEBGL2.R8UI,
		r16i: WEBGL2.R16I,
		r16ui: WEBGL2.R16UI,
		r32i: WEBGL2.R32I,
		r32ui: WEBGL2.R32UI,
		rg8i: WEBGL2.RG8I,
		rg8ui: WEBGL2.RG8UI,
		rg16i: WEBGL2.RG16I,
		rg16ui: WEBGL2.RG16UI,
		rg32i: WEBGL2.RG32I,
		rg32ui: WEBGL2.RG32UI,
		r8_snorm: WEBGL2.R8_SNORM,
		rg8_snorm: WEBGL2.RG8_SNORM,
		rgb8_snorm: WEBGL2.RGB8_SNORM,
		rgba8_snorm: WEBGL2.RGBA8_SNORM,
		rgb10_a2ui: WEBGL2.RGB10_A2UI,
		texture_immutable_format: WEBGL2.TEXTURE_IMMUTABLE_FORMAT,
		texture_immutable_levels: WEBGL2.TEXTURE_IMMUTABLE_LEVELS,
	},
	uniform_types:
	{
		float_vec2: WEBGL2.FLOAT_VEC2,
		float_vec3: WEBGL2.FLOAT_VEC3,
		float_vec4: WEBGL2.FLOAT_VEC4,
		int_vec2: WEBGL2.INT_VEC2,
		int_vec3: WEBGL2.INT_VEC3,
		int_vec4: WEBGL2.INT_VEC4,
		bool: WEBGL2.BOOL,
		bool_vec2: WEBGL2.BOOL_VEC2,
		bool_vec3: WEBGL2.BOOL_VEC3,
		bool_vec4: WEBGL2.BOOL_VEC4,
		float_mat2: WEBGL2.FLOAT_MAT2,
		float_mat3: WEBGL2.FLOAT_MAT3,
		float_mat4: WEBGL2.FLOAT_MAT4,
		sampler_2d: WEBGL2.SAMPLER_2D,
		sampler_cube: WEBGL2.SAMPLER_CUBE,
	},

	frame_render_buffers:
	{
		framebuffer: WEBGL2.FRAMEBUFFER,
		renderbuffer: WEBGL2.RENDERBUFFER,
		rgba4: WEBGL2.RGBA4,
		rgb5_a1: WEBGL2.RGB5_A1,
		rgb565: WEBGL2.RGB565,
		depth_component16: WEBGL2.DEPTH_COMPONENT16,
		stencil_index: WEBGL2.STENCIL_INDEX,
		stencil_index8: WEBGL2.STENCIL_INDEX8,
		depth_stencil: WEBGL2.DEPTH_STENCIL,
		renderbuffer_width: WEBGL2.RENDERBUFFER_WIDTH,
		renderbuffer_height: WEBGL2.RENDERBUFFER_HEIGHT,
		renderbuffer_internal_format: WEBGL2.RENDERBUFFER_INTERNAL_FORMAT,
		renderbuffer_red_size: WEBGL2.RENDERBUFFER_RED_SIZE,
		renderbuffer_green_size: WEBGL2.RENDERBUFFER_GREEN_SIZE,
		renderbuffer_blue_size: WEBGL2.RENDERBUFFER_BLUE_SIZE,
		renderbuffer_alpha_size: WEBGL2.RENDERBUFFER_ALPHA_SIZE,
		renderbuffer_depth_size: WEBGL2.RENDERBUFFER_DEPTH_SIZE,
		renderbuffer_stencil_size: WEBGL2.RENDERBUFFER_STENCIL_SIZE,
		framebuffer_attachment_object_type: WEBGL2.FRAMEBUFFER_ATTACHMENT_OBJECT_TYPE,
		framebuffer_attachment_object_name: WEBGL2.FRAMEBUFFER_ATTACHMENT_OBJECT_NAME,
		framebuffer_attachment_texture_level: WEBGL2.FRAMEBUFFER_ATTACHMENT_TEXTURE_LEVEL,
		framebuffer_attachment_texture_cube_map_face: WEBGL2.FRAMEBUFFER_ATTACHMENT_TEXTURE_CUBE_MAP_FACE,
		color_attachment0: WEBGL2.COLOR_ATTACHMENT0,
		depth_attachment: WEBGL2.DEPTH_ATTACHMENT,
		stencil_attachment: WEBGL2.STENCIL_ATTACHMENT,
		depth_stencil_attachment: WEBGL2.DEPTH_STENCIL_ATTACHMENT,
		none: WEBGL2.NONE,
		framebuffer_complete: WEBGL2.FRAMEBUFFER_COMPLETE,
		framebuffer_incomplete_attachment: WEBGL2.FRAMEBUFFER_INCOMPLETE_ATTACHMENT,
		framebuffer_incomplete_missing_attachment: WEBGL2.FRAMEBUFFER_INCOMPLETE_MISSING_ATTACHMENT,
		framebuffer_incomplete_dimensions: WEBGL2.FRAMEBUFFER_INCOMPLETE_DIMENSIONS,
		framebuffer_unsupported: WEBGL2.FRAMEBUFFER_UNSUPPORTED,
		framebuffer_binding: WEBGL2.FRAMEBUFFER_BINDING,
		renderbuffer_binding: WEBGL2.RENDERBUFFER_BINDING,
		max_renderbuffer_size: WEBGL2.MAX_RENDERBUFFER_SIZE,
		invalid_framebuffer_operation: WEBGL2.INVALID_FRAMEBUFFER_OPERATION,
		framebuffer_attachment_color_encoding: WEBGL2.FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING,
		framebuffer_attachment_component_type: WEBGL2.FRAMEBUFFER_ATTACHMENT_COMPONENT_TYPE,
		framebuffer_attachment_red_size: WEBGL2.FRAMEBUFFER_ATTACHMENT_RED_SIZE,
		framebuffer_attachment_green_size: WEBGL2.FRAMEBUFFER_ATTACHMENT_GREEN_SIZE,
		framebuffer_attachment_blue_size: WEBGL2.FRAMEBUFFER_ATTACHMENT_BLUE_SIZE,
		framebuffer_attachment_alpha_size: WEBGL2.FRAMEBUFFER_ATTACHMENT_ALPHA_SIZE,
		framebuffer_attachment_depth_size: WEBGL2.FRAMEBUFFER_ATTACHMENT_DEPTH_SIZE,
		framebuffer_attachment_stencil_size: WEBGL2.FRAMEBUFFER_ATTACHMENT_STENCIL_SIZE,
		framebuffer_default: WEBGL2.FRAMEBUFFER_DEFAULT,
		depth24_stencil8: WEBGL2.DEPTH24_STENCIL8,
		draw_framebuffer_binding: WEBGL2.DRAW_FRAMEBUFFER_BINDING,
		read_framebuffer: WEBGL2.READ_FRAMEBUFFER,
		draw_framebuffer: WEBGL2.DRAW_FRAMEBUFFER,
		read_framebuffer_binding: WEBGL2.READ_FRAMEBUFFER_BINDING,
		renderbuffer_samples: WEBGL2.RENDERBUFFER_SAMPLES,
		framebuffer_attachment_texture_layer: WEBGL2.FRAMEBUFFER_ATTACHMENT_TEXTURE_LAYER,
		framebuffer_incomplete_multisample: WEBGL2.FRAMEBUFFER_INCOMPLETE_MULTISAMPLE,
	},
	query:
	{
		current_query: WEBGL2.CURRENT_QUERY,
		query_result: WEBGL2.QUERY_RESULT,
		query_result_available: WEBGL2.QUERY_RESULT_AVAILABLE,
		any_samples_passed: WEBGL2.ANY_SAMPLES_PASSED,
		any_samples_passed_conservative: WEBGL2.ANY_SAMPLES_PASSED_CONSERVATIVE,
	},
	draw_buffers:
	{
		max_draw_buffers: WEBGL2.MAX_DRAW_BUFFERS,
		draw_buffer0: WEBGL2.DRAW_BUFFER0,
		draw_buffer1: WEBGL2.DRAW_BUFFER1,
		draw_buffer2: WEBGL2.DRAW_BUFFER2,
		draw_buffer3: WEBGL2.DRAW_BUFFER3,
		draw_buffer4: WEBGL2.DRAW_BUFFER4,
		draw_buffer5: WEBGL2.DRAW_BUFFER5,
		draw_buffer6: WEBGL2.DRAW_BUFFER6,
		draw_buffer7: WEBGL2.DRAW_BUFFER7,
		draw_buffer8: WEBGL2.DRAW_BUFFER8,
		draw_buffer9: WEBGL2.DRAW_BUFFER9,
		draw_buffer10: WEBGL2.DRAW_BUFFER10,
		draw_buffer11: WEBGL2.DRAW_BUFFER11,
		draw_buffer12: WEBGL2.DRAW_BUFFER12,
		draw_buffer13: WEBGL2.DRAW_BUFFER13,
		draw_buffer14: WEBGL2.DRAW_BUFFER14,
		draw_buffer15: WEBGL2.DRAW_BUFFER15,
		max_color_attachments: WEBGL2.MAX_COLOR_ATTACHMENTS,
		color_attachment1: WEBGL2.COLOR_ATTACHMENT1,
		color_attachment2: WEBGL2.COLOR_ATTACHMENT2,
		color_attachment3: WEBGL2.COLOR_ATTACHMENT3,
		color_attachment4: WEBGL2.COLOR_ATTACHMENT4,
		color_attachment5: WEBGL2.COLOR_ATTACHMENT5,
		color_attachment6: WEBGL2.COLOR_ATTACHMENT6,
		color_attachment7: WEBGL2.COLOR_ATTACHMENT7,
		color_attachment8: WEBGL2.COLOR_ATTACHMENT8,
		color_attachment9: WEBGL2.COLOR_ATTACHMENT9,
		color_attachment10: WEBGL2.COLOR_ATTACHMENT10,
		color_attachment11: WEBGL2.COLOR_ATTACHMENT11,
		color_attachment12: WEBGL2.COLOR_ATTACHMENT12,
		color_attachment13: WEBGL2.COLOR_ATTACHMENT13,
		color_attachment14: WEBGL2.COLOR_ATTACHMENT14,
		color_attachment15: WEBGL2.COLOR_ATTACHMENT15,
	},
	samplers:
	{
		sampler_3d: WEBGL2.SAMPLER_3D,
		sampler_2d_shadow: WEBGL2.SAMPLER_2D_SHADOW,
		sampler_2d_array: WEBGL2.SAMPLER_2D_ARRAY,
		sampler_2d_array_shadow: WEBGL2.SAMPLER_2D_ARRAY_SHADOW,
		sampler_cube_shadow: WEBGL2.SAMPLER_CUBE_SHADOW,
		int_sampler_2d: WEBGL2.INT_SAMPLER_2D,
		int_sampler_3d: WEBGL2.INT_SAMPLER_3D,
		int_sampler_cube: WEBGL2.INT_SAMPLER_CUBE,
		int_sampler_2d_array: WEBGL2.INT_SAMPLER_2D_ARRAY,
		unsigned_int_sampler_2d: WEBGL2.UNSIGNED_INT_SAMPLER_2D,
		unsigned_int_sampler_3d: WEBGL2.UNSIGNED_INT_SAMPLER_3D,
		unsigned_int_sampler_cube: WEBGL2.UNSIGNED_INT_SAMPLER_CUBE,
		unsigned_int_sampler_2d_array: WEBGL2.UNSIGNED_INT_SAMPLER_2D_ARRAY,
		max_samples: WEBGL2.MAX_SAMPLES,
		sampler_binding: WEBGL2.SAMPLER_BINDING,
	},
	transform_feedback:
	{
		transform_feedback_buffer_mode: WEBGL2.TRANSFORM_FEEDBACK_BUFFER_MODE,
		max_transform_feedback_separate_components: WEBGL2.MAX_TRANSFORM_FEEDBACK_SEPARATE_COMPONENTS,
		transform_feedback_varyings: WEBGL2.TRANSFORM_FEEDBACK_VARYINGS,
		transform_feedback_buffer_start: WEBGL2.TRANSFORM_FEEDBACK_BUFFER_START,
		transform_feedback_buffer_size: WEBGL2.TRANSFORM_FEEDBACK_BUFFER_SIZE,
		transform_feedback_primitives_written: WEBGL2.TRANSFORM_FEEDBACK_PRIMITIVES_WRITTEN,
		max_transform_feedback_interleaved_components: WEBGL2.MAX_TRANSFORM_FEEDBACK_INTERLEAVED_COMPONENTS,
		max_transform_feedback_separate_attribs: WEBGL2.MAX_TRANSFORM_FEEDBACK_SEPARATE_ATTRIBS,
		interleaved_attribs: WEBGL2.INTERLEAVED_ATTRIBS,
		separate_attribs: WEBGL2.SEPARATE_ATTRIBS,
		transform_feedback_buffer: WEBGL2.TRANSFORM_FEEDBACK_BUFFER,
		transform_feedback_buffer_binding: WEBGL2.TRANSFORM_FEEDBACK_BUFFER_BINDING,
		transform_feedback: WEBGL2.TRANSFORM_FEEDBACK,
		transform_feedback_paused: WEBGL2.TRANSFORM_FEEDBACK_PAUSED,
		transform_feedback_active: WEBGL2.TRANSFORM_FEEDBACK_ACTIVE,
		transform_feedback_binding: WEBGL2.TRANSFORM_FEEDBACK_BINDING,
	},
	uniforms:
	{
		uniform_buffer: WEBGL2.UNIFORM_BUFFER,
		uniform_buffer_binding: WEBGL2.UNIFORM_BUFFER_BINDING,
		uniform_buffer_start: WEBGL2.UNIFORM_BUFFER_START,
		uniform_buffer_size: WEBGL2.UNIFORM_BUFFER_SIZE,
		max_vertex_uniform_blocks: WEBGL2.MAX_VERTEX_UNIFORM_BLOCKS,
		max_fragment_uniform_blocks: WEBGL2.MAX_FRAGMENT_UNIFORM_BLOCKS,
		max_combined_uniform_blocks: WEBGL2.MAX_COMBINED_UNIFORM_BLOCKS,
		max_uniform_buffer_bindings: WEBGL2.MAX_UNIFORM_BUFFER_BINDINGS,
		max_uniform_block_size: WEBGL2.MAX_UNIFORM_BLOCK_SIZE,
		max_combined_vertex_uniform_components: WEBGL2.MAX_COMBINED_VERTEX_UNIFORM_COMPONENTS,
		max_combined_fragment_uniform_components: WEBGL2.MAX_COMBINED_FRAGMENT_UNIFORM_COMPONENTS,
		uniform_buffer_offset_alignment: WEBGL2.UNIFORM_BUFFER_OFFSET_ALIGNMENT,
		active_uniform_blocks: WEBGL2.ACTIVE_UNIFORM_BLOCKS,
		uniform_type: WEBGL2.UNIFORM_TYPE,
		uniform_size: WEBGL2.UNIFORM_SIZE,
		uniform_block_index: WEBGL2.UNIFORM_BLOCK_INDEX,
		uniform_offset: WEBGL2.UNIFORM_OFFSET,
		uniform_array_stride: WEBGL2.UNIFORM_ARRAY_STRIDE,
		uniform_matrix_stride: WEBGL2.UNIFORM_MATRIX_STRIDE,
		uniform_is_row_major: WEBGL2.UNIFORM_IS_ROW_MAJOR,
		uniform_block_binding: WEBGL2.UNIFORM_BLOCK_BINDING,
		uniform_block_data_size: WEBGL2.UNIFORM_BLOCK_DATA_SIZE,
		uniform_block_active_uniforms: WEBGL2.UNIFORM_BLOCK_ACTIVE_UNIFORMS,
		uniform_block_active_uniform_indices: WEBGL2.UNIFORM_BLOCK_ACTIVE_UNIFORM_INDICES,
		uniform_block_referenced_by_vertex_shader: WEBGL2.UNIFORM_BLOCK_REFERENCED_BY_VERTEX_SHADER,
		uniform_block_referenced_by_fragment_shader: WEBGL2.UNIFORM_BLOCK_REFERENCED_BY_FRAGMENT_SHADER,
	},
	sync:
	{
		object_type: WEBGL2.OBJECT_TYPE,
		sync_condition: WEBGL2.SYNC_CONDITION,
		sync_status: WEBGL2.SYNC_STATUS,
		sync_flags: WEBGL2.SYNC_FLAGS,
		sync_fence: WEBGL2.SYNC_FENCE,
		sync_gpu_commands_complete: WEBGL2.SYNC_GPU_COMMANDS_COMPLETE,
		unsignaled: WEBGL2.UNSIGNALED,
		signaled: WEBGL2.SIGNALED,
		already_signaled: WEBGL2.ALREADY_SIGNALED,
		timeout_expired: WEBGL2.TIMEOUT_EXPIRED,
		condition_satisfied: WEBGL2.CONDITION_SATISFIED,
		wait_failed: WEBGL2.WAIT_FAILED,
		sync_flush_commands_bit: WEBGL2.SYNC_FLUSH_COMMANDS_BIT,
	},
	misc:
	{
		color: WEBGL2.COLOR,
		depth: WEBGL2.DEPTH,
		stencil: WEBGL2.STENCIL,
		min: WEBGL2.MIN,
		max: WEBGL2.MAX,
		depth_component24: WEBGL2.DEPTH_COMPONENT24,
		stream_read: WEBGL2.STREAM_READ,
		stream_copy: WEBGL2.STREAM_COPY,
		static_read: WEBGL2.STATIC_READ,
		static_copy: WEBGL2.STATIC_COPY,
		dynamic_read: WEBGL2.DYNAMIC_READ,
		dynamic_copy: WEBGL2.DYNAMIC_COPY,
		depth_component32f: WEBGL2.DEPTH_COMPONENT32F,
		depth32f_stencil8: WEBGL2.DEPTH32F_STENCIL8,
		invalid_index: WEBGL2.INVALID_INDEX,
		timeout_ignored: WEBGL2.TIMEOUT_IGNORED,
		max_client_wait_timeout_webgl: WEBGL2.MAX_CLIENT_WAIT_TIMEOUT_WEBGL,
	},
	angle_instanced_arrays:
	{
		vertex_attrib_array_divisor_angle: WEBGL2.VERTEX_ATTRIB_ARRAY_DIVISOR_ANGLE,
	},
	webgl_debug_renderer_info:
	{
		unmasked_vendor_webgl: WEBGL2.UNMASKED_VENDOR_WEBGL,
		unmasked_renderer_webgl: WEBGL2.UNMASKED_RENDERER_WEBGL,
	},
	ext_texture_filter_anisotropic:
	{
		max_texture_max_anisotropy_ext: WEBGL2.MAX_TEXTURE_MAX_ANISOTROPY_EXT,
		texture_max_anisotropy_ext: WEBGL2.TEXTURE_MAX_ANISOTROPY_EXT,
	},
	webgl_compressed_texture_s3tc:
	{
		compressed_rgb_s3tc_dxt1_ext: WEBGL2.COMPRESSED_RGB_S3TC_DXT1_EXT,
		compressed_rgba_s3tc_dxt1_ext: WEBGL2.COMPRESSED_RGBA_S3TC_DXT1_EXT,
		compressed_rgba_s3tc_dxt3_ext: WEBGL2.COMPRESSED_RGBA_S3TC_DXT3_EXT,
		compressed_rgba_s3tc_dxt5_ext: WEBGL2.COMPRESSED_RGBA_S3TC_DXT5_EXT,
	},
	webgl_compressed_texture_etc:
	{
		compressed_r11_eac: WEBGL2.COMPRESSED_R11_EAC,
		compressed_signed_r11_eac: WEBGL2.COMPRESSED_SIGNED_R11_EAC,
		compressed_rg11_eac: WEBGL2.COMPRESSED_RG11_EAC,
		compressed_signed_rg11_eac: WEBGL2.COMPRESSED_SIGNED_RG11_EAC,
		compressed_rgb8_etc2: WEBGL2.COMPRESSED_RGB8_ETC2,
		compressed_rgba8_etc2_eac: WEBGL2.COMPRESSED_RGBA8_ETC2_EAC,
		compressed_srgb8_etc2: WEBGL2.COMPRESSED_SRGB8_ETC2,
		compressed_srgb8_alpha8_etc2_eac: WEBGL2.COMPRESSED_SRGB8_ALPHA8_ETC2_EAC,
		compressed_rgb8_punchthrough_alpha1_etc2: WEBGL2.COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2,
		compressed_srgb8_punchthrough_alpha1_etc2: WEBGL2.COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2,
	},
	webgl_compressed_texture_pvrtc:
	{
		compressed_rgb_pvrtc_4bppv1_img: WEBGL2.COMPRESSED_RGB_PVRTC_4BPPV1_IMG,
		compressed_rgba_pvrtc_4bppv1_img: WEBGL2.COMPRESSED_RGBA_PVRTC_4BPPV1_IMG,
		compressed_rgb_pvrtc_2bppv1_img: WEBGL2.COMPRESSED_RGB_PVRTC_2BPPV1_IMG,
		compressed_rgba_pvrtc_2bppv1_img: WEBGL2.COMPRESSED_RGBA_PVRTC_2BPPV1_IMG,
	},
	webgl_compressed_texture_etc1:
	{
		compressed_rgb_etc1_webgl: WEBGL2.COMPRESSED_RGB_ETC1_WEBGL,
	},
	webgl_compressed_texture_atc:
	{
		compressed_rgb_atc_webgl: WEBGL2.COMPRESSED_RGB_ATC_WEBGL,
		compressed_rgba_atc_explicit_alpha_webgl: WEBGL2.COMPRESSED_RGBA_ATC_EXPLICIT_ALPHA_WEBGL,
		compressed_rgba_atc_interpolated_alpha_webgl: WEBGL2.COMPRESSED_RGBA_ATC_INTERPOLATED_ALPHA_WEBGL,
	},
	webgl_depth_texture:
	{
		unsigned_int_24_8_webgl: WEBGL2.UNSIGNED_INT_24_8_WEBGL,
	},
	oes_texture_half_float:
	{
		half_float_oes: WEBGL2.HALF_FLOAT_OES,
	},
	webgl_color_buffer_float:
	{
		rgba32f_ext: WEBGL2.RGBA32F_EXT,
		rgb32f_ext: WEBGL2.RGB32F_EXT,
		framebuffer_attachment_component_type_ext: WEBGL2.FRAMEBUFFER_ATTACHMENT_COMPONENT_TYPE_EXT,
		unsigned_normalized_ext: WEBGL2.UNSIGNED_NORMALIZED_EXT,
	},
	ext_blend_minmax:
	{
		min_ext: WEBGL2.MIN_EXT,
		max_ext: WEBGL2.MAX_EXT,
	},
	ext_srgb:
	{
		srgb_ext: WEBGL2.SRGB_EXT,
		srgb_alpha_ext: WEBGL2.SRGB_ALPHA_EXT,
		srgb8_alpha8_ext: WEBGL2.SRGB8_ALPHA8_EXT,
		framebuffer_attachment_color_encoding_ext: WEBGL2.FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING_EXT,
	},
	oes_standard_derivatives:
	{
		fragment_shader_derivative_hint_oes: WEBGL2.FRAGMENT_SHADER_DERIVATIVE_HINT_OES,
	},
	webgl_draw_buffers:
	{
		color_attachment0_webgl: WEBGL2.COLOR_ATTACHMENT0_WEBGL,
		color_attachment1_webgl: WEBGL2.COLOR_ATTACHMENT1_WEBGL,
		color_attachment2_webgl: WEBGL2.COLOR_ATTACHMENT2_WEBGL,
		color_attachment3_webgl: WEBGL2.COLOR_ATTACHMENT3_WEBGL,
		color_attachment4_webgl: WEBGL2.COLOR_ATTACHMENT4_WEBGL,
		color_attachment5_webgl: WEBGL2.COLOR_ATTACHMENT5_WEBGL,
		color_attachment6_webgl: WEBGL2.COLOR_ATTACHMENT6_WEBGL,
		color_attachment7_webgl: WEBGL2.COLOR_ATTACHMENT7_WEBGL,
		color_attachment8_webgl: WEBGL2.COLOR_ATTACHMENT8_WEBGL,
		color_attachment9_webgl: WEBGL2.COLOR_ATTACHMENT9_WEBGL,
		color_attachment10_webgl: WEBGL2.COLOR_ATTACHMENT10_WEBGL,
		color_attachment11_webgl: WEBGL2.COLOR_ATTACHMENT11_WEBGL,
		color_attachment12_webgl: WEBGL2.COLOR_ATTACHMENT12_WEBGL,
		color_attachment13_webgl: WEBGL2.COLOR_ATTACHMENT13_WEBGL,
		color_attachment14_webgl: WEBGL2.COLOR_ATTACHMENT14_WEBGL,
		color_attachment15_webgl: WEBGL2.COLOR_ATTACHMENT15_WEBGL,
		draw_buffer0_webgl: WEBGL2.DRAW_BUFFER0_WEBGL,
		draw_buffer1_webgl: WEBGL2.DRAW_BUFFER1_WEBGL,
		draw_buffer2_webgl: WEBGL2.DRAW_BUFFER2_WEBGL,
		draw_buffer3_webgl: WEBGL2.DRAW_BUFFER3_WEBGL,
		draw_buffer4_webgl: WEBGL2.DRAW_BUFFER4_WEBGL,
		draw_buffer5_webgl: WEBGL2.DRAW_BUFFER5_WEBGL,
		draw_buffer6_webgl: WEBGL2.DRAW_BUFFER6_WEBGL,
		draw_buffer7_webgl: WEBGL2.DRAW_BUFFER7_WEBGL,
		draw_buffer8_webgl: WEBGL2.DRAW_BUFFER8_WEBGL,
		draw_buffer9_webgl: WEBGL2.DRAW_BUFFER9_WEBGL,
		draw_buffer10_webgl: WEBGL2.DRAW_BUFFER10_WEBGL,
		draw_buffer11_webgl: WEBGL2.DRAW_BUFFER11_WEBGL,
		draw_buffer12_webgl: WEBGL2.DRAW_BUFFER12_WEBGL,
		draw_buffer13_webgl: WEBGL2.DRAW_BUFFER13_WEBGL,
		draw_buffer14_webgl: WEBGL2.DRAW_BUFFER14_WEBGL,
		draw_buffer15_webgl: WEBGL2.DRAW_BUFFER15_WEBGL,
		max_color_attachments_webgl: WEBGL2.MAX_COLOR_ATTACHMENTS_WEBGL,
		max_draw_buffers_webgl: WEBGL2.MAX_DRAW_BUFFERS_WEBGL,
	},
	oes_vertex_array_object:
	{
		vertex_array_binding_oes: WEBGL2.VERTEX_ARRAY_BINDING_OES,
	},
	ext_disjoint_timer_query:
	{
		query_counter_bits_ext: WEBGL2.QUERY_COUNTER_BITS_EXT,
		current_query_ext: WEBGL2.CURRENT_QUERY_EXT,
		query_result_ext: WEBGL2.QUERY_RESULT_EXT,
		query_result_available_ext: WEBGL2.QUERY_RESULT_AVAILABLE_EXT,
		time_elapsed_ext: WEBGL2.TIME_ELAPSED_EXT,
		timestamp_ext: WEBGL2.TIMESTAMP_EXT,
		gpu_disjoint_ext: WEBGL2.GPU_DISJOINT_EXT,
	}
};
